### PR TITLE
Rely on backend build path and bootstrap metadata to easily build from source

### LINKED ISSUE #1644: Tests failing to run with BackendUnavailable since pip 19
It seems that since the release of pip 19, Setuptools' own tests fail to run thus:

```
setuptools fix_889_and_non-ascii_in_setup.cfg_take_2 $ tox -e py27                                                                                                                       
py27 create: /Users/jaraco/code/main/setuptools/.tox/py27
py27 installdeps: -rtests/requirements.txt
py27 develop-inst: /Users/jaraco/code/main/setuptools
ERROR: invocation failed (exit code 2), logfile: /Users/jaraco/code/main/setuptools/.tox/py27/log/py27-2.log
ERROR: actionid: py27
msg: developpkg
cmdargs: '/Users/jaraco/code/main/setuptools/.tox/py27/bin/pip install --exists-action w -e /Users/jaraco/code/main/setuptools'

DEPRECATION: Python 2.7 will reach the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 won't be maintained after that date. A future version of pip will drop support for Python 2.7.
Obtaining file:///Users/jaraco/code/main/setuptools
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
Exception:
Traceback (most recent call last):
  File "/Users/jaraco/code/main/setuptools/.tox/py27/lib/python2.7/site-packages/pip/_internal/cli/base_command.py", line 176, in main
    status = self.run(options, args)
  File "/Users/jaraco/code/main/setuptools/.tox/py27/lib/python2.7/site-packages/pip/_internal/commands/install.py", line 315, in run
    resolver.resolve(requirement_set)
  File "/Users/jaraco/code/main/setuptools/.tox/py27/lib/python2.7/site-packages/pip/_internal/resolve.py", line 131, in resolve
    self._resolve_one(requirement_set, req)
  File "/Users/jaraco/code/main/setuptools/.tox/py27/lib/python2.7/site-packages/pip/_internal/resolve.py", line 294, in _resolve_one
    abstract_dist = self._get_abstract_dist_for(req_to_install)
  File "/Users/jaraco/code/main/setuptools/.tox/py27/lib/python2.7/site-packages/pip/_internal/resolve.py", line 226, in _get_abstract_dist_for
    req, self.require_hashes, self.use_user_site, self.finder,
  File "/Users/jaraco/code/main/setuptools/.tox/py27/lib/python2.7/site-packages/pip/_internal/operations/prepare.py", line 382, in prepare_editable_requirement
    abstract_dist.prep_for_dist(finder, self.build_isolation)
  File "/Users/jaraco/code/main/setuptools/.tox/py27/lib/python2.7/site-packages/pip/_internal/operations/prepare.py", line 149, in prep_for_dist
    reqs = self.req.pep517_backend.get_requires_for_build_wheel()
  File "/Users/jaraco/code/main/setuptools/.tox/py27/lib/python2.7/site-packages/pip/_vendor/pep517/wrappers.py", line 71, in get_requires_for_build_wheel
    'config_settings': config_settings
  File "/Users/jaraco/code/main/setuptools/.tox/py27/lib/python2.7/site-packages/pip/_vendor/pep517/wrappers.py", line 162, in _call_hook
    raise BackendUnavailable
BackendUnavailable

py27 installed: DEPRECATION: Python 2.7 will reach the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 won't be maintained after that date. A future version of pip will drop support for Python 2.7.,apipkg==1.5,atomicwrites==1.2.1,attrs==18.2.0,configparser==3.7.1,contextlib2==0.5.5,coverage==4.5.2,enum34==1.1.6,execnet==1.5.0,flake8==3.6.0,funcsigs==1.0.2,futures==3.2.0,importlib-metadata==0.8,mccabe==0.6.1,mock==2.0.0,more-itertools==5.0.0,path.py==11.5.0,pathlib2==2.3.3,pbr==5.1.1,pluggy==0.8.1,py==1.7.0,pycodestyle==2.4.0,pyflakes==2.0.0,pytest==3.10.1,pytest-cov==2.6.1,pytest-fixture-config==1.4.0,pytest-flake8==1.0.3,pytest-shutil==1.4.0,pytest-virtualenv==1.4.0,scandir==1.9.0,six==1.12.0,termcolor==1.1.0,virtualenv==16.3.0,zipp==0.3.3
________________________________________________________________________________________ summary ________________________________________________________________________________________
ERROR:   py27: InvocationError for command /Users/jaraco/code/main/setuptools/.tox/py27/bin/pip install --exists-action w -e /Users/jaraco/code/main/setuptools (see /Users/jaraco/code/main/setuptools/.tox/py27/log/py27-2.log) (exited with code 2)
```

I suspect this is related to #1642 if not just another manifestation of the same issue.

### PR DESCRIPTION
**Pull request by jaraco**
Fixes #1644.

### CODE REVIEW

File: setuptools/tests/test_virtualenv.py [hunk 1]
```diff
@@ -69,7 +69,9 @@ def _get_pip_versions():
         'pip==9.0.3',
         'pip==10.0.1',
         'pip==18.1',
-        'pip==19.0.1',
+        # fails due to pypa/pip#6599
+        # 'pip==19.3.1',
```
Comment (webknjaz) [inline setuptools/tests/test_virtualenv.py:73]: Maybe xfail it instead?
```suggestion
        pytest.mark.xfail('pip==19.3.1', reason='fails due to pypa/pip#6599'),
```

File: pyproject.toml [hunk 2]
```diff
@@ -1,6 +1,5 @@
 [build-system]
 requires = [
-    "setuptools >= 40.8",
     "wheel",
 ]
 build-backend = "setuptools.build_meta"
```
Comment (webknjaz) [inline pyproject.toml:5]: Have you considered having maybe a completely separate backend?

File: CHANGES.rst [hunk 3]
```diff
@@ -5,7 +5,7 @@ v53.1.0
 Changes
 ^^^^^^^
 * #1937: Preserved case-sensitivity of keys in setup.cfg so that entry point names are case-sensitive. Changed sensitivity of configparser. NOTE: Any projects relying on case-insensitivity will need to adapt to accept the original case as published. -- by :user:`melissa-kun-li`
-* #2573: Fixed error in uploading a Sphinx doc with the :code:`upload_docs` command. An html builder will be used. 
+* #2573: Fixed error in uploading a Sphinx doc with the :code:`upload_docs` command. An html builder will be used.
```
Comment (webknjaz) [inline CHANGES.rst:8]: Looks like this should go to a separate PR

File: setuptools/tests/test_virtualenv.py [hunk 4]
```diff
@@ -69,7 +69,9 @@ def _get_pip_versions():
         'pip==9.0.3',
         'pip==10.0.1',
         'pip==18.1',
-        'pip==19.0.1',
+        # fails due to pypa/pip#6599
+        pytest.mark.xfail('pip==19.3.1', reason='fails due to pypa/pip#6599'),
```
Comment (webknjaz) [inline setuptools/tests/test_virtualenv.py:73]: Sorry, I posted the previous comment from mobile and messed up the syntax. I think it should be
```suggestion
        pytest.param(
            'pip==19.3.1',
            marks=pytest.mark.xfail(reason='fails due to pypa/pip#6599'),
        ),
```

File: pyproject.toml (hunk 2 above)
Comment (jaraco) [reply pyproject.toml:5]: I have. Maybe something that vendors all of the dependencies or just hard-codes the behavior especially for setuptools. It's not out of consideration yet.

File: CHANGES.rst (hunk 3 above)
Comment (jaraco) [reply CHANGES.rst:8]: Yep. Was an incidental change. Fixed in 1a549254c.

Comment (jaraco) [reply CHANGES.rst]: And rebased.

File: setuptools/tests/test_virtualenv.py (hunk 4 above)
Comment (jaraco) [reply setuptools/tests/test_virtualenv.py]: Dang. So when I rebased and pushed for the other comment, I overwrote the patch. No matter. I'll get this change in.

Comment (jaraco) [reply setuptools/tests/test_virtualenv.py]: Fixed in 13ff0ba92.

File: pyproject.toml (hunk 2 above)
Comment (webknjaz) [reply pyproject.toml:5]: Yep, that's exactly what I had in my mind.
